refactor: split `MetalsLanguageServer` into language server and language service

kpodsiad opened the pull request:
Previously, in `MetalsLanguageServer` there were multiple services which initialization has been delayed (`= _`) because project workspace or client config was needed. This is not needed as former `MetalsLanguageServer` can be instantiated only after receiving `Initialize` request from the client:

```mermaid
sequenceDiagram
participant C as LSP client
participant S as LSP server

C->>S: Initialize request <br> (workspace + client config)
S->>C: Initiliazed response
Note over C,S: Client must not send any <br> request until receiving <br> Initialized response
C-->>S: Initialized notification
Note over C,S: The Initialized notification is send <br> before client can send any other <br> request or notification to the server
```

Based on that knowledge I extracted those parts as `LanguageServer` trait. Similarly, I extracted `TextDocumentService`, `WorkspaceService` and `MetalsService` into traits. Traits are needed to create default, dummy implementation. This is due to how [JsonDelegate](https://github.com/eclipse/lsp4j/blob/main/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/services/JsonDelegate.java) works - in runtime class is being processed and if method annotated with delegate returns null then no methods will be registered. In runtime, at the beginning we can't provide valid implementation of those services (because we delay init after `Initialize` request) - hence traits with default, throwing implementation. 
Why I didn't used [lsp4j's services](https://github.com/eclipse/lsp4j/tree/main/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/services) - well sometimes return types of methods differ and I didn't dive if there's difference in how lsp4j handles Unit vs Future[Unit]

todo: 
- [x] pass workspace and client config as constructor args in old `MetalsLanguageServer`, move initialization logic to the body
- [x] reduce mutability when possible (and easy)
- [x] rename old `MetalsLanguageServer` to e.g. `MetalsLanguageServer`. Any better ideas? I want to rename that just before merge to make conflicts easier to resolve
- [x] fix tests:
- [x] fix`StatusBarSuite` -  there was an issue with `ClientConfiguration`
- [x] fix `CallHierarchyLspSuite` - split tests to not call "initialize" twice
- [x] fix some issues with thread pools cancelation - add scaladoc to make intentions clearer
- [x] fix integration tests
- [x] investigate issues with RejectedExecutionException which are visible in logs. Although, as @tgodzik pointed, in tests server is being cancelled after (to be precise before in beforeEach) each test and thread pools are being shutdown after each suite. 
- [x] double check cancelation logic, make sure that there is no resource leak

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -102,14 +98,15 @@ import io.undertow.server.HttpServerExchange
 import org.eclipse.{lsp4j => l}
 
+// todo rename to MetalsLspService
 class MetalsLanguageServer(
```
Or MetalsServerWorkspace ?

kpodsiad commented:
Ok, CI is quite ok right now, macos is failing but I believe those are flaky tests. One thing which still worries me - there are `org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.nio.channels.AsynchronousCloseException` and `java.util.concurrent.RejectedExecutionException: Task Future() rejected from` exceptions in logs. They might be cause by my changes, but it seems that they're not affecting correctness of Metals.

tgodzik commented:
> Ok, CI is quite ok right now, macos is failing but I believe those are flaky tests. One thing which still worries me - there are `org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.nio.channels.AsynchronousCloseException` and `java.util.concurrent.RejectedExecutionException: Task Future() rejected from` exceptions in logs. They might be cause by my changes, but it seems that they're not affecting correctness of Metals.

I think it's because we shutdown the server mutiple times in the tests and the errors/warnings are legit, but not an issue.

kpodsiad commented:
I've squashed commits, only meaningful ones remained. PR is ready for review.

kpodsiad commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -171,20 +209,22 @@ class MetalsLanguageServer(
+  private val sourceMapper = SourceMapper(
+    buildTargets,
+    buffers,
+    () => workspace,
```
`() => workspace` can be changed to just `workspace`

ckipp01 replied:
Just to be sure, this is because the workspace is already known right?

kpodsiad replied:
yup

kpodsiad commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -241,650 +281,569 @@ class MetalsLanguageServer(
+
+  private val documentSymbolProvider = new DocumentSymbolProvider(
+    trees,
+    Some(initializeParams).supportsHierarchicalDocumentSymbols,
```
Extension method can operate on plain `InitializeParams` rather than Option of it

ckipp01 replied:
Are you sure? I didn't see extensions methods in `MetalsEnrichments` for this. Or do you mean we should switch that?

kpodsiad replied:
It is `XtensionClientCapabilities` which operates on `initializeParams: Option[l.InitializeParams]`

ckipp01 replied:
Yes, exactly, that's what I mean, it operates on an `Option`. Does your original comment mean you'd like to change this to operate directly on `l.InitializeParams` instead of the `Option`?

kpodsiad commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -241,650 +281,569 @@ class MetalsLanguageServer(
+      compilations,
+      statusBar,
+      diagnostics,
+      () => tables,
```
`() => tables` => just `tables`

kpodsiad commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,272 @@
+    scribe.error("Unexpected error initializing server", e)
+  }.asJava
+
+  override def shutdown(): CompletableFuture[Unit] = serverState match {
+    case ServerState.Initialized(server) =>
+      server.shutdown()
+    case _ =>
+      Future.unit.asJava
+  }
+
+  override def exit(): Unit = serverState match {
+    case ServerState.ShuttingDown(server) =>
+      server.exit()
+    case _ => ()
+  }
```
Maybe it's worth to add logging form each branch, just in case?

ckipp01 replied:
Are you referring to the shutdown/exits? If so sure.

kpodsiad replied:
yes, both cases (happy and error paths) for both shutdown and exit

kpodsiad commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,272 @@
+
+  override val getScalaService: ScalaLspService = metalsService
+
+  @deprecated
+  def getOldMetalsLanguageServer: MetalsLspService = serverState match {
+    case ServerState.Initialized(server) => server
+    case _ => throw new IllegalStateException("Server is not initialized")
+  }
```
Used only in tests, I didn't want to alter tests more than it was needed to fix failing ones. Additional work can be done in follow up of this PR

ckipp01 replied:
I'd say create an issue right away for this then to outline what would need to be changed and to remove this after.

ckipp01 commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,272 @@
+
+  private val cancelables = new MutableCancelable()
+  private val isCancelled = new AtomicBoolean(false)
+  private val isLanguageClientConnected = new AtomicBoolean(false)
```
Again, excuse the naive question, but related to your comment above. In this situation for example we only use `set` and `get`, but are still using an `AtomicBoolean` over a `@volatile`. How come?

kpodsiad replied:
good point, I'm not being consistent. As I wrote earlier, it's probably easier to switch to `Atomic*` and just forget about `@volatile`

ckipp01 commented on metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:
```diff
@@ -0,0 +1,206 @@
+ * [[org.eclipse.lsp4j.jsonrpc.services.JsonDelegate]]. Value returned by
+ * method with that annotation is used to determine which methods are supported
+ * by the server. We can't initialize it to null because then no lsp methods would be
+ * supported. Instead, we initialize it to a dummy instance which is then replaced.
```
Bear with me, I'll probably ask a bunch of these just t make sure I'm tracking, but in case the dummy on is the `new ScalaLspService {}` which will then later be replaced by the `MetalsLspService` as the _real_ one right?

kpodsiad replied:
yes, you're correct.
The thing is, at runtime (I'm not sure at exactly which point though), probably betwenn those lines https://github.com/kpodsiad/metals/blob/908c0c1e199626abcf3375d96555772568e794ab/metals/src/main/scala/scala/meta/metals/Main.scala#L59-L70 lsp4j processes given server which was set in `setLocalService` method. If method with `JsonDelegate` annotation exists, that method is being called. If it returns null processing is aborted, if not them processing is continued on class which is being returned. 

The tricky part - now we delay creation of MetalsLspService so we cannot provide working instance from the very beginning. We can't return null because lsp4j processing won't process `ScalaLspService` methods and our MetalsLanguageServer will "see" only basic init/shutdown endpoints*

* - it is also possible that I misunderstood some parts of how lsp4j works, but after I reached current state I was happy enough to see things working.

Answering your questions in this thread so we can keep discussion in one place:

`ScalaLspService` isn't just a Dummy instance, it's an interface which describes endpoints. That interface is being later on implemented by MetalsLspService. What may be con fusing is fact, that `ScalaLspService` mixes two roles:
- it's interface
- it provides default implementation for methods so it is possible to do just `new ScalaLspService {}`

Default implementations are needed because of later point. Otherwise all of those methods had to be implemented

Are things more clear now? Maybe scaladocs needs some improvement?

ckipp01 replied:
Yea this helps, thanks.

kpodsiad replied:
Should I modify scaladoc to make things crystal clear?

ckipp01 replied:
I think more than a Scaladoc we should try and outline the entire thing in the design doc.

ckipp01 commented on metals/src/main/scala/scala/meta/metals/lsp/LanguageServer.scala:
```diff
@@ -0,0 +1,49 @@
+  def getScalaService: ScalaLspService
+}
+
+trait ScalaLspService
+    extends TextDocumentService
+    with WorkspaceService
+    with MetalsService
```
Let's add some scaladocs here just to make it super clear that this is the Dummy instance that doesn't do anything.

ckipp01 commented on metals/src/main/scala/scala/meta/metals/lsp/MetalsService.scala:
```diff
@@ -0,0 +1,79 @@
+ * implemented by Metals.
+ *
+ * Each method has a default implementation which throws UnsupportedOperationException. Throwing in this context is ok because:
+ * - the default implementation should never be called, because the method is always overridden in [[scala.meta.internal.metals.MetalsLanguageServer]]
```
Out of curiosity, do these even need a default implementation then? Couldn't they just all be abstract?

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -102,14 +98,25 @@ import io.undertow.server.HttpServerExchange
-import org.eclipse.lsp4j.jsonrpc.services.JsonRequest
 import org.eclipse.{lsp4j => l}
 
+// todo rename to MetalsLspService
```
I think `MetalsLspService` is fine. Could we finish up the scaladocs here as well?

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/StatusBar.scala:
```diff
@@ -36,7 +36,7 @@ import org.eclipse.lsp4j.MessageType
  *   as long as the attached `Future[_]` is not completed.
  */
 final class StatusBar(
-    client: () => MetalsLanguageClient,
```
Just to be sure, we did this before because we wanted to ensure we got the configured client, but in this scenario, we _always_ have the configured client right?

kpodsiad replied:
Exactly. Previously, language client was obtained and set after creating MetalsLanguageServer - https://github.com/scalameta/metals/blob/3cbb04ef03d2d390b228744aa5c5824aa6a8274e/metals/src/main/scala/scala/meta/metals/Main.scala#L68-L70.

Now, with MetalsLanguageServer being refactored and new MetalsLspService, the later one is created with workspace and proper client passed in so there is no need for `() => ` workarounds

kpodsiad replied:
[Here](https://github.com/kpodsiad/metals/blob/908c0c1e199626abcf3375d96555772568e794ab/metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala#L176-L192), we're creating MetalsService

ckipp01 reviewed (commented):
Really really fantastic refactoring @kpodsiad. I love seeing all of the setting to `_` removed. This really cleans some stuff up. I just mainly have a couple questions/nits.

Since quite a bit of stuff is changing at the moment, would you also be able to run through https://github.com/scalameta/metals/blob/main/architecture.md and maybe update / expand the relevant parts about the changes that were made here?

kpodsiad commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,272 @@
+
+  // AtomicReference's set and get have the same semantics as volatile variables.
+  // because we only use get and set, there is no need to use AtomicReference.
+  @volatile
```
Only saves a few keystrokes when getting and setting value -  `=` vs `get()/set()`. On the other hand I wrote 2 lines of comment to describe why volatile is good enough 😅 
Maybe it's simpler to just use `AtomicReference` and _not think of_ anything

ckipp01 replied:
ha, I don't have strong feelings on it, was just curious. Either way, I think consistency is important. So if we go with `@volatile`, then we should do it for all that we can. If not, then yea, let's just stick with the `AtomicReference`.

kpodsiad commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,272 @@
+ * @param ec
+ *   execution context for futures.
+ * @param sh
+ *   scheduled executor service for scheduling tasks.
```
yes we can, but I'd delay it a bit. I think we can do much better job when it comes to configuration - https://github.com/scalameta/metals/issues/4789

ckipp01 replied:
👍🏼 as long as you plan on following up with those other issues, then no worries. If you aren't going to get to them right away, then I'd say add them now.

kpodsiad replied:
I had this in mind, but I also see that @tanishiking assigned @kasiaMarek in https://github.com/scalameta/metals/issues/4789. It seems that definitely someone will handle this ;)

kpodsiad commented:
@ckipp01 I think that I addressed your comments either directly or by creating follow up tasks and placing todos in code. I'd like to merge this PR as soon as it is possible to test it thoroughly before next release.

ckipp01 commented:
> @ckipp01 I think that I addressed your comments either directly or by creating follow up tasks and placing todos in code. I'd like to merge this PR as soon as it is possible to test it thoroughly before next release.

I think we want to release before we merge this no? It's been quite some time before we released, so I'd actually recommend trying to finish up https://github.com/scalameta/metals/pull/4740 as quickly as we can, cut a release, and then get this merged. I wouldn't try to merge this before that release since there is already a ton of stuff in there.

@tgodzik is there anything I can help with for the release or are there blockers yet?

kpodsiad commented:
> I think we want to release before we merge this no? It's been quite some time before we released, so I'd actually recommend trying to finish up #4740 as quickly as we can, cut a release, and then get this merged. I wouldn't try to merge this before that release since there is already a ton of stuff in there.

Yes, that's the idea, to merge it after current release.

kpodsiad commented:
I renamed old `MetalsLanguageServer` to `MetalsLanguageService`. I'll update `architecture.md` in separate PR.
@ckipp01 @tgodzik can I have an approval on this one? ;)

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:
```diff
@@ -39,6 +39,9 @@ import scala.meta.io.AbsolutePath
 import org.eclipse.lsp4j.Position
 
+// todo https://github.com/scalameta/metals/issues/4788
+// clean () =>, use plain values
```
This one should be fairly simple right? Are you sure you don't want to just tackle this right away? Or do you still want to merge and then send in a follow-up for this? The reason I ask is because you already dit this for `tables` in here. So its odd that you did one but not the others.

kpodsiad replied:
One should be really careful when replacing `() =>`, in most cases this will be straighforward change, but sometimes `() =>` was used to capture mutable state. I already made that mistake and it took some time to find culprit. 

> The reason I ask is because you already dit this for tables in here. So its odd that you did one but not the others.

Well, doing it for tables made me realize how much such stuff lives in codebase. I saw those 2 classes, but probably there are more. So if I did it for Tables, Indexer then why not for others? ;) This PR is already big and I strongly believe in small (if possible), with well-defined scope, short-living ones. That's why I'm pushing all of those things into smaller, auxiliary tasks. They're relatively easy so maybe even someone else will do them, who knows :D

ckipp01 reviewed (changes requested):
The only blocker at this point for me is making sure that https://github.com/scalameta/metals/blob/main/architecture.md is updated. I do have a few small nits, but if you plan on following up with the other issues, then no worries.

- [ ] Update architecture.md

ckipp01 commented:
> I'll update architecture.md in separate PR.

pfff, I totally missed this in your last comment.  😬

ckipp01 reviewed (approved):
LGTM from me, but I think you should also wait for another ✅ from @tgodzik since this is a huge change.

tgodzik commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,267 @@
+        case _ if !isLanguageClientConnected.get =>
+          Future
+            .failed(
+              new IllegalStateException("Language client wasn't plugged!")
```
```suggestion
              new IllegalStateException("Language client wasn't connected!")
```
?

tgodzik commented on metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:
```diff
@@ -0,0 +1,267 @@
+  // todo https://github.com/scalameta/metals/issues/4785
+  @deprecated
+  def getOldMetalsLanguageServer: MetalsLspService = serverState.get match {
+    case ServerState.Initialized(server) => server
```
```suggestion
    case ServerState.Initialized(service) => service
```

tgodzik commented on metals/src/main/scala/scala/meta/metals/lsp/TextDocumentService.scala:
```diff
@@ -0,0 +1,171 @@
+ * - the default implementation should never be called, because the method is always overridden in [[scala.meta.internal.metals.MetalsLanguageServer]]
+ * - lsp4j wraps each method in try catch and returns an error response to the client
+ */
+trait TextDocumentService {
```
So this not implements all possible methods from LSP spec? Even if it isn't yet implemented by Metals?

kpodsiad replied:
No, it contains only those which are implemented by Metals

tgodzik commented on metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:
```diff
@@ -0,0 +1,206 @@
+    var underlying: ScalaLspService
+) extends ScalaLspService {
+
+  override def didOpen(
```
There are only the ones implemented by Metals?

So now when adding a new functionality, you would need to add the method in 3 places? Could you add a comment about it in each of the classes we might need to add it, as noticable as possible? :sweat_smile: 

There was already a bit of issues with the DelegatingMetalsClient etc. (that is still needed, no?)

kpodsiad replied:
Changed in [dff67d4](https://github.com/scalameta/metals/pull/4776/commits/dff67d4fb2d241a03bf340a62114eb44231a0a8c). Now I pass null as initial value to delegating service (still fine as it's not called until initialize and correct value is set before). I got rid of default implementations so now adding new endpoint in trait will cause compilation errors in both DelegatingService and in MetalsLspService (which are exactly 2 places which needs to be changed)

tgodzik reviewed (changes requested):
This is a very epic and courageous PR ! Very well done! I added some comments, mostly minor stuff and you will probably need to rebase on the semantic tokens PR, but that should really be just one method to add.

kpodsiad commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:
```diff
@@ -241,650 +282,569 @@ class MetalsLanguageServer(
-          new ExecuteCommandOptions(
-            (ServerCommands.allIds ++ codeActionProvider.allActionCommandsIds).toList.asJava
-          )
+          new ExecuteCommandOptions(ServerCommands.allIds.toList.asJava)
```
you're right, I can't recall why I did that

kpodsiad commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:
```diff
@@ -241,650 +282,569 @@ class MetalsLanguageServer(
-  def initialize(
-      params: InitializeParams
-  ): CompletableFuture[InitializeResult] = {
+  def initialize(): CompletableFuture[InitializeResult] = {
```
definitely, 95% of this method is just setting LSP capabilities

kpodsiad commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:
```diff
@@ -2765,7 +2678,7 @@ class MetalsLanguageServer(
   private def syncUserconfiguration(): Future[Unit] = {
     val supportsConfiguration = for {
-      params <- initializeParams
+      params <- Some(initializeParams)
```
yeah, it's leftover

tgodzik reviewed (approved):
LGTM! This looks great! Go ahead with the merge :D
